samples/drivers/soc_flash_nrf/src/main.c

```diff
@@ -93,27 +86,32 @@ int main(void)
 			  flash_params_get_erase_cap(&flash_params) & FLASH_ERASE_C_EXPLICIT,
 			  TEST_PARTITION_OFFSET, FLASH_PAGE_SIZE);
 
-	printf("\nTest 2: Internal storage write (word array 1)\n");
-	for (i = 0U; i < ARRAY_SIZE(buf_array_1); i++) {
-		offset = TEST_PARTITION_OFFSET + (i << 2);
-		printf("   Attempted to write %x at 0x%x\n", buf_array_1[i],
-				offset);
-		if (flash_write(flash_dev, offset, &buf_array_1[i],
-					sizeof(uint32_t)) != 0) {
-			printf("   Write failed!\n");
+	printf("	Write block size: %u\n", flash_params.write_block_size);
+	write_cycles = ARRAY_SIZE(test_data) / flash_params.write_block_size;
+	printf("	Required write cycles for given data and memory: %u\n", write_cycles);
+
+	printf("\nTest 2: Internal storage write\n");
+	for (i = 0U; i < write_cycles; i++) {
+		offset = TEST_PARTITION_OFFSET + i * flash_params.write_block_size;
+		printf("	Writing %u data bytes to 0x%x\n", flash_params.write_block_size,
+		       offset);
+		if (flash_write(flash_dev, offset, &test_data[i], flash_params.write_block_size) !=
+		    0) {
+			printf("	Write failed!\n");
 			return 0;
 		}
-		printf("   Attempted to read 0x%x\n", offset);
-		if (flash_read(flash_dev, offset, &buf_word,
-					sizeof(uint32_t)) != 0) {
-			printf("   Read failed!\n");
+		printf("	Reading %u data bytes from 0x%x\n", flash_params.write_block_size,
+		       offset);
+		if (flash_read(flash_dev, offset, &read_buffer[i], flash_params.write_block_size) !=
+		    0) {
+			printf("	Read failed!\n");
 			return 0;
 		}
-		printf("   Data read: %x\n", buf_word);
-		if (buf_array_1[i] == buf_word) {
-			printf("   Data read matches data written. Good!\n");
+
+		if (memcmp(test_data, read_buffer, flash_params.write_block_size)) {
+			printf("	Data read does not match data written!\n");
 		} else {
-			printf("   Data read does not match data written!\n");
+			printf("	Data read matches data written. Good!\n");
 		}
```

Comment:
This loop should get own function as it is nearly the same as between lines 136-154. The only difference is the "de-alignment" to offset, which is +1, which can be parameter to the function.